Objects/bytearrayobject.c

```diff
@@ -844,8 +844,23 @@ bytearray___init___impl(PyByteArrayObject *self, PyObject *arg,
         return -1;
     }
 
-    /* XXX Optimize this if the arguments is a list, tuple */
+    if (PyList_CheckExact(arg) || PyTuple_CheckExact(arg)) {
+        Py_ssize_t size = PySequence_Fast_GET_SIZE(arg);
+        if (PyByteArray_Resize((PyObject *)self, size) < 0) {
+            return -1;
+        }
+        PyObject **items = PySequence_Fast_ITEMS(arg);
 
+        for (Py_ssize_t i = 0; i < size; i++) {
+            int value;
+            int rc = _getbytevalue(items[i], &value);
```

Comment:
I've found a path through `_getbytevalue()` that may invoke Python code. That Python code could mutate the argument if it's a list, invalidating `items`, causing a read of freed memory.

The path goes as follows:

- `_getbytevalue`
- `PyLong_AsLongAndOverflow`
- `_PyNumber_Index`

One of the items must be a Python class implementing `__index__`. In that `__index__` method you could mutate the list argument to `bytes()`.

(There's also a much longer path that goes through the warning printed from `_PyNumber_Index`.)

To avoid this, you could check for `PyLong_CheckExact(items[i])` before calling `_getbytevalue()`, and if it isn't, abandon the fast path.

Reply:
This is a good catch, I found a similar bug in the existing code: https://bugs.python.org/issue46997

Reply:
Fixed